### ISSUE TITLE
Generated file cleanup

### ISSUE THREAD

**Issue by leonardt** [IntelLabs/ParallelAccelerator.jl#4]
When using MPI but a shared .julia, multiple julia processes will use `deps/generated` to store generated files, causing a filename conflict.  I've resolved this locally by attaching `getpid()` to the generated filenames and keeping track of generated files and cleaning them up with `atexit()`.  

This presents a problem when Julia doesn't exit cleanly, leaving generated files in `deps/generated` that will not be cleaned up.  Currently this isn't really an issue because cgen will overwrite these files the next time ParallelAccelerator is used, however this would not happen with the above `getpid()` change because the process id will have changed.  This leads to the question of how we should deal with zombie generated files.

One solution is to move the generated files to `/tmp` so the OS will take care of it on reboot by using `tempdir()`

Thoughts?

**Comment by ninegua** [comment]
Agree. I suggest tmp dir and/or random filename (see Unix command `mktemp` or `tempfile`) when user doesn't have PROSPECT_DEV_MODE turned on.

**Comment by leonardt** [comment]
sounds good, Julia has some built in bindings for `mktemp` and `tempfile` so I'll use those

**Comment by ehsantn** [comment]
But the generated files should be easy to find and look at for the developers. Maybe we need another switch for when we need to save them?

**Comment by leonardt** [comment]
Yes, we can leave the current behavior for PROSPECT_DEV_MODE and then use tmp dir by default

**Comment by lkuper** [comment]
@leonardt I'm now seeing errors like:

```
julia> include("black-scholes/black-scholes.jl")
iterations = 10000000
ERROR: LoadError: SystemError: opening file /tmp/tmpdCuNBF/cgen_output1.cpp: No such file or directory
 in open at ./iostream.jl:90
 in open at ./iostream.jl:102
 in __writec#470__ at /home/lkuper/.julia/v0.4/ParallelAccelerator/src/cgen.jl:2218
 in __writec#469__ at /home/lkuper/.julia/v0.4/ParallelAccelerator/src/cgen.jl:2211
 in toCGen at /home/lkuper/.julia/v0.4/ParallelAccelerator/src/driver.jl:148
 in processFuncCall at /home/lkuper/.julia/v0.4/CompilerTools/src/OptFramework.jl:338
 in blackscholes at /home/lkuper/.julia/v0.4/CompilerTools/src/OptFramework.jl:400
 in main at /mnt/home/lkuper/.julia/v0.4/ParallelAccelerator/examples/black-scholes/black-scholes.jl:92
 in include at ./boot.jl:261
 in include_from_node1 at ./loading.jl:304
while loading /mnt/home/lkuper/.julia/v0.4/ParallelAccelerator/examples/black-scholes/black-scholes.jl, in expression starting on line 101
```

Could that be a result of this change?

**Comment by leonardt** [comment]
I saw that, do you have precompilation turned on?  You can check to see if there is a ParallelAccelerator.ji in `ls ~/.julia/lib/v0.4`.  Basically it creates a tempfile to store generated files each run, but then removes this file `atexit`.  But with precompilation it inlines the tempfile name so a new one isn't created every time the package is loaded, explaining the directory does not exist error.  This should be solvable using `__init__()` though and I'm looking into that

**Comment by leonardt** [comment]
@lkuper I was able to reproduce with compilation turned on and pushed a fix with 4dfea9be5c564b65bedebf6a8106db55ba8cd3bf.  Basically it checks if the directory exists at runtime and if not generates a new one.  Let me know if you still see the issue

**Comment by lkuper** [comment]
@leonardt I have your patch now and I have the same issue, exactly the same as before.
